Implement some Sys stubs

ryanartecona opened the pull request:
This implements enough of the yet-unimplemented runtime primitives & stdlib that the cmdliner library can be compiled unmodified by bucklescript & the result run in node.

Specifically, I've implemented these:
- `Sys.command`
- `Sys.file_exists`
- `Sys.is_directory`
- `Sys.remove` via primitive `caml_sys_remove`
- `Pervasives.(open_out, open_out_bin, open_out_gen)` via primitives `caml_sys_open` and `caml_ml_open_descriptor_out`
- `Pervasives.(open_in, input, input_char, input_line, really_input, really_input_string, close_in)` via primities `caml_ml_open_descriptor_in`, `caml_ml_input`, `caml_ml_input_char`, and `caml_ml_input_scan_line`.

I've also made a small change to `Sys.argv` s.t. the executable name takes up only 1 slot, as in OCaml (and compiled langs, typically), rather than 2 slots (i.e. `["/usr/local/bin/node", "script.js"]`, as is typical in node's `process.argv`.

Most of the new functionality is in the 2 new runtime modules `caml_sys_system` and `caml_sys_fs` (better name suggestions welcome). The idea is that `caml_sys_system` depends on the node `child_process` module and `caml_sys_fs` depends on node's `fs`. If a library uses functionality which transitively depends on `Sys.command`, then the compiled JS modules will transitively require `child_process`, but if it doesn't, they won't. Same goes for `Sys.file_exists` and friends with `fs`. These runtime modules won't work in a browser context, of course, but if a user tries to webpack or rollup them, then they should get an error when the bundle module resolver can't find those modules or detects that they're internal node modules.

---
Edit: Added details for in_channel & input primitive implementations

bobzhang commented:
@ryanartecona this is quite impressive work.
I have another idea that instead of polyfill the underlying c primitive, how hard is to patch sys.ml (https://github.com/BuckleScript/bucklescript/blob/master/jscomp/stdlib/sys.ml) directly with the same sys.mli interface?

ryanartecona commented:
@bobzhang thanks!

I'm happy to try that if you think it would work. I was worried that adding bs externals which depend on node's `fs` and `child_process` modules would mean `fs` and `child_process` would become transitive dependencies of any bs code that uses anything in Sys. That would complicate things for folks using webpack or rollup to bundle browser js. Is that accurate? Or would bucklescript only include module dependencies for externals that get used (using cross-module DCE or something)?

I guess I can test this out easily enough myself.

bobzhang commented:
@ryanartecona I grepped usage of Sys module in standard library:
Sys.argv (in arg module, I think it is okay to introduce node env?)
Sys.max_array_length
Sys.max_string_length
Sys.word_size
Sys.getenv 
(in Hashtbl - can be patched to remove it)
(in Filename - I wish we can rewrite it on top of NodeJS path)
Sys.os_type

bobzhang commented:
Another direction we can explore is that by stripping stdlib out of the compiler, we introduce a conditional compilation MACRO, so that nodejs users and browser users can share the same code base while making great use of its own platform

ryanartecona commented:
> So if your rewrite of Sys module is functional (require or not does not have side effect),
> we can short-cut such usages above without depending on Sys module

I'm not sure I follow. What mechanism would allow us to do that? And what do you mean by "short-cut" here?

I believe all my changes should be functional. One possible future exception might be the primitive `caml_ml_out_channels_list`, which is essentially a global mutable list of all open out_channels used by `Pervasives.flush_all`. Right now I've left that to just a static `[stdout; stderr]` which is technically incorrect but I haven't needed to change it for my own purposes.

> Another direction we can explore is that by stripping stdlib out of the compiler, we introduce a conditional compilation MACRO, so that nodejs users and browser users can share the same code base while making great use of its own platform

This probably deserves more thought and more visibility than just this PR.

FWIW, this PR shouldn't negatively affect anything in Sys that was already working in a browser context. It implements some stuff for a node context that previously just threw that "not yet implemented" runtime error, but in a browser context the new stuff will just start being a module resolution error (at bundle or require time) instead of that "not implemented" runtime error. On that note, can we merge this when it's ready and take the bigger question of how to best treat browser/node stdlib compatibility to a new issue?

ryanartecona commented:
Update: I've added implementations `Pervasives.in_channel` and most of the related primitives and updated the description to match.

I see now that the `Caml_sys_fs` runtime module is pulled into the compiled JS of stdlib modules (e.g. Pervasives, Digest, Filename, Scanf). Not sure how I missed that before. That's definitely not mergeable since it would break the compiled JS for everyone targeting a browser context, which is probably most users at this point.

I need to figure out a way that this implementation can exist without breaking things for browser targets.

bobzhang commented:
Hi, I feel really sorry but our focus shifted these years and the top priority is to optimize js experience. Your effort is much appreciated.
